relay-server/src/actors/envelopes.rs

```diff
@@ -688,120 +721,252 @@ impl EnvelopeProcessor {
+            relay_log::trace!("fixing session timestamp to {}", session.timestamp);
+            session.timestamp = session.started;
+            changed = true;
+        }
 
-            // Log the timestamp delay for all sessions after clock drift correction.
-            let session_delay = received - session.timestamp;
-            if session_delay > SignedDuration::minutes(1) {
-                metric!(
-                    timer(RelayTimers::TimestampDelay) = session_delay.to_std().unwrap(),
-                    category = "session",
-                );
-            }
+        // Log the timestamp delay for all sessions after clock drift correction.
+        let session_delay = *received - session.timestamp;
+        if session_delay > SignedDuration::minutes(1) {
+            metric!(
+                timer(RelayTimers::TimestampDelay) = session_delay.to_std().unwrap(),
+                category = "session",
+            );
+        }
 
-            if (received - session.started) > max_age || (received - session.timestamp) > max_age {
-                relay_log::trace!("skipping session older than {} days", max_age.num_days());
+        // Validate timestamps
+        for t in &[&session.timestamp, &session.started] {
+            if !self.is_valid_session_timestamp(received, t) {
                 return false;
             }
+        }
 
-            let max_future = SignedDuration::seconds(self.config.max_secs_in_future());
-            if (session.started - received) > max_future
-                || (session.timestamp - received) > max_future
-            {
-                relay_log::trace!(
-                    "skipping session more than {}s in the future",
-                    max_future.num_seconds()
-                );
-                return false;
+        // Validate attributes
+        match self.validate_attributes(client_addr, &mut session.attributes) {
+            Err(_) => return false,
+            Ok(changed_attributes) => {
+                changed |= changed_attributes;
             }
+        }
 
-            let release = &session.attributes.release;
-            if let Err(e) = protocol::validate_release(release) {
-                relay_log::trace!("skipping session with invalid release '{}': {}", release, e);
+        // Extract metrics
+        if metrics_extraction_enabled && !item.metrics_extracted() {
+            extract_session_metrics(&session.attributes, &session, extracted_metrics);
+            item.set_metrics_extracted(true);
+        }
+
+        if changed {
+            let json_string = match serde_json::to_string(&session) {
+                Ok(json) => json,
+                Err(err) => {
+                    relay_log::error!("failed to serialize session: {}", LogError(&err));
+                    return false;
+                }
+            };
+
+            item.set_payload(ContentType::Json, json_string);
+        }
+
+        true
+    }
+
+    fn process_session_aggregates(
```

Comment:
It might make sense to create a common interface for `Session` and `SessionAggregate` (similar to but different from `SessionLike`) instead, so that we only have to define `process_session` once.